language-plutus-core/generators/Language/PlutusCore/Generators/Test.hs

```diff
@@ -1,18 +1,116 @@
+-- | This module defines various types and functions useful for testing.
+
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
 module Language.PlutusCore.Generators.Test
-    ( propEvaluate
+    ( TypeEvalCheckError (..)
+    , TypeEvalCheckResult (..)
+    , TypeEvalCheckM
+    , typeEvalCheckBy
+    , unsafeTypeEvalCheck
+    , sampleTermValue
+    , sampleProgramValueGolden
+    , propEvaluate
     ) where
 
 import           Language.PlutusCore
 import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Evaluation.CkMachine
 import           Language.PlutusCore.Evaluation.Result
-import           Language.PlutusCore.Generators
+import           Language.PlutusCore.Generators.Interesting
+import           Language.PlutusCore.Generators.Internal.TypedBuiltinGen
+import           Language.PlutusCore.Generators.Internal.Utils
 import           Language.PlutusCore.Pretty
-import           PlutusPrelude                         (PrettyConfigIgnore (..))
+import           PlutusPrelude                                           hiding (hoist, (</>))
 
 import           Control.Monad.Except
 import           Control.Monad.Morph
-import           Data.Foldable
-import           Hedgehog                              hiding (Size, Var, eval)
+import qualified Data.Text.IO                                            as Text
+import           Data.Traversable
+import           Hedgehog                                                hiding (Size, Var, eval)
+import           System.FilePath                                         ((</>))
+
+{- Note [Type-eval checking]
+We generate terms along with values they are supposed to evaluate to. Before evaluating a term,
+we type check it. Then we evaluate the term and check whether the expected result matches with
+the actual one. Thus "type-eval checking".
+-}
+
+-- | The type of errors that can occur during type-eval checking.
+data TypeEvalCheckError
+    = TypeEvalCheckErrorIllFormed (Error ())
+    | TypeEvalCheckErrorIllEvaled (Value TyName Name ()) (Value TyName Name ())
```

Comment:
Can you add haddock for which is expected vs actual here, otherwise it's easy to get confused.

Reply:
Haddock doesn't support comments like that, I think. I usually do it like that:

```haskell
type Expected a = a
type Actual a = a

TypeEvalCheckErrorIllEvaled (Expected (Value TyName Name ())) (Actual (Value TyName Name ()))
```

If that's fine for your taste, I can do this.

Reply:
I'd settle  for a comment on the type itself.

Reply:
I commented the constructor.